lib/Dialect/SV/Transforms/HWExportModuleHierarchy.cpp

```diff
@@ -0,0 +1,135 @@
+//===- HWExportModuleHierarchy.cpp - Export Module Hierarchy ----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Export the module and instance hierarchy information to JSON. This pass
+// stores the JSON information in an sv.verbatim op with the output_file
+// attribute set.
+//
+// Note that two hierarchies are exported to separate files. One is rooted at
+// the main module as indicated by the firrtl.mainModule attribute, while the
+// other is rooted at the design under test, as indicated by the DesignUnderTest
+// attribute.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "mlir/IR/Builders.h"
+
+using namespace circt;
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+struct HWExportModuleHierarchyPass
+    : public sv::HWExportModuleHierarchyBase<HWExportModuleHierarchyPass> {
+  void runOnOperation() override;
+};
+
+/// Recursively print the module hierarchy as serialized as JSON.
+static void printHierarchy(hw::InstanceOp &inst, SymbolTable &symbolTable,
+                           llvm::raw_ostream &os) {
+  os << "{"
+     << "\"instance_name\":\"" << inst.instanceName() << "\","
+     << "\"module_name\":\"" << inst.moduleName() << "\","
+     << "\"instances\":[";
+
+  auto moduleOp =
+      symbolTable.lookup<hw::HWModuleOp>(inst.moduleNameAttr().getValue());
+
+  // Only recurse on module ops, not extern or generated ops, whose internal are
+  // opaque.
+  if (moduleOp) {
+    bool firstIteration = true;
+    for (auto op : moduleOp.getOps<hw::InstanceOp>()) {
+      if (firstIteration)
+        firstIteration = false;
+      else
+        os << ",";
+      printHierarchy(op, symbolTable, os);
+    }
+  }
+
+  os << "]}";
```

Comment:
This can all be simplified by using the LLVM JSON library.  I think this is as straightforward as pulling in the include:

```c++
#include "llvm/Support/JSON.h"
```

And then replacing the body here with something like:
```suggestion
  json::OStream J(os);

  auto moduleOp =
      symbolTable.lookup<hw::HWModuleOp>(inst.moduleNameAttr().getValue());

  J.object([&] {
    J.attribute("instance_name", inst.instanceName());
    J.attribute("module_name", inst.moduleName());
    J.attributeArray("instances", [&] {
      for (auto op : moduleOp.getOps<hw::InstanceOp>())
        printHierarchy(op, symbolTable, os);
    });
  });
```

Reply:
Thanks for pointing this out! I should have figured that LLVM would have its own JSON library. Will definitely update this code to make use of it.